### ISSUE TITLE
[Feature Request] Allow update() to create & use state & temp variables

### ISSUE THREAD

**Issue by Nostromos** [astashov/liftosaur#104]
**Which problem is this feature request solving?**
While many of the common progression methods are based on results at the end of an exercise, others require progression *during* an exercise (Mike Mathews version of double progression is a great example). 

**Describe the solution you'd like**
- Allow `update()` to create and set state and temporary variables, along with reading them OR
- Allow 'progress()` to trigger after set completion.

**Describe alternatives you've considered**
I've attempted to use a combination of progress and update but without the ability for update to write state variables, the code is convoluted and no true connection can be made between the two of them.

**Can you submit a pull request?**
Yes, if I get the ok from creator/maintainer. Not sure if this is something you'd want to do, Anton. 

--- 
Here's the backstory from my post on discord:

> I'm doing the Bigger, Leaner, Stronger 5-day routine and trying to program in his (Mike Matthews) version of double progression:
> 
> Let’s say you’re squatting in the 4-to-6-rep range, and on your first (or second) hard set in the workout, you get 6 reps of 225 pounds. You then immediately add 10 pounds to the bar (not in your next workout), rest a few minutes, and get 4 or 5 reps on your next hard set. Great! The progression has succeeded, and you now work with 235 pounds in your current and future workouts until you can squat it for one hard set of 6 reps. Then, you’d move up in weight again, and so on. 
> 
> What if you only get 2 or 3 reps with the new, heavier weight before your form starts breaking down? What if the progression doesn’t succeed? In this case, you should drop the weight back to the original, lighter load (225 pounds in our example) and work there until you can get 6 reps for two hard sets (in the same workout). Then, you should immediately move up to the heavier weight again on your next hard set (even if that’s in your next workout) and try again. If you do that and still can’t get at least 4 reps, go back to the lighter weight and work with it until you can do three hard sets of 6 reps (in the same workout). 
> 
> At this point, your progression should succeed. Also, if you get 6 reps in your first set after moving up in weight, then you get to move up again! And what should you do if you get 6 reps on your third and final hard set for an exercise in your workout? You should increase the weight on your first hard set of your next performance of that exercise.
> 
> 
> So, my takeaway here is:
> If you hit max reps on 1 set, go up. If you hit min reps on next set, stay with that weight.
> If the next set doesn't hit min reps, go down until you hit max reps for 2 sets, then...
> Try the progression again. Go up. If you can't hit min reps, go back down until you can hit max reps for 3 sets.
> Then go up and stay there. This assumes you can do max reps for 3 sets of lower weight and min reps for 3 sets of higher weight. I've never had the problem that I can do 3 sets at max reps of lower weight but not hit minimum at higher weight, so I'm not gonna solve for it yet.
> 
> More simply:
> 1 failure to successfully progress needs 2 sets at lower weight with max reps
> 2 failures to successfully progress needs 3 sets at lower weight
> if you hit 3 sets at max, progress and stay no matter what.
> 
> Does this custom progression look right? The issue I'm having is that the progression isn't triggering during the workout, only after I finish the day.
>
> 
> ```js
> custom(needed: 1, successes: 0, increment: 10lb) {~
>   for (var.i in completedReps) {
>     if (completedReps[var.i] >= max(reps[var.i])) {
>       state.successes += 1
>       if (state.needed == state.successes) {
>         weights += state.increment
>         state.successes = 0 // Reset successes after increasing the weight
>       }
>       if (state.needed == 3 && state.successes == 3) {
>         state.needed = 1
>         state.successes = 0
>       }
>     } else if (completedReps[var.i] < min(reps[var.i])) {
>       weights -= state.increment
>       state.needed += 1
>       if (state.needed > 3) {
>         state.needed = 3 // Cap needed at 3
>       }
>       state.successes = 0 // Reset successes after decreasing the weight
>     } 
> ~}
> ```

**Comment by kevinmessiaen** [comment]
Yes this is a feature would really be beneficial, a simple example is from Starting Strength. If you fail to achieve the 15 reps in your 3 sets, then you should do singles/doubles to match the failed reps.

Making `progress` to be triggered at each set will mess up existing programs so it's not a good idea. But adding another method callback can be a simple and efficient solution:

```js
custom / used: none / 3x5 / progress: ...startingStrength / onSetDone(initialReps: 0) {~
  if (state.initialReps == 0) {
    state.initalReps = reps
  }

  if (completedSet == numberOfSets && completedReps <= state.initalReps) {
    numberOfSets += 1
    // TODO: Set reps of last set to `completedReps - state.initalReps`
  }
~}
```
